Sample/EventsVersioning/EventsVersioning.Tests/ExplicitSerialization/ShoppingCart.cs

```diff
@@ -0,0 +1,413 @@
+    {
+    }
+
+    public int CompareTo(LocalDateTime? other) => other != null ? Value.CompareTo(other.Value) : -1;
+
+
+    public static LocalDateTime Parse(DateTimeOffset value) => new(value);
+}
+
+public record Money(
+    Amount Amount,
+    Currency Currency
+);
+
+public class Price: StronglyTypedValue<Money>
+{
+    public Price(Money value): base(value) { }
+
+    public static Price Parse(Money value)
+    {
+        if (!value.Amount.IsPositive)
+            throw new ArgumentOutOfRangeException(nameof(value), "Price cannot be negative");
+
+        return new Price(value);
+    }
+}
+
+public abstract record ShoppingCartEvent
+{
+    public record ShoppingCartOpened(
+        ShoppingCartId ShoppingCartId,
+        ClientId ClientId
+    ): ShoppingCartEvent;
+
+    public record ProductItemAddedToShoppingCart(
+        ShoppingCartId ShoppingCartId,
+        PricedProductItem ProductItem
+    ): ShoppingCartEvent;
+
+    public record ProductItemRemovedFromShoppingCart(
+        ShoppingCartId ShoppingCartId,
+        PricedProductItem ProductItem
+    ): ShoppingCartEvent;
+
+    public record ShoppingCartConfirmed(
+        ShoppingCartId ShoppingCartId,
+        LocalDateTime ConfirmedAt
+    ): ShoppingCartEvent;
+
+    public record ShoppingCartCanceled(
+        ShoppingCartId ShoppingCartId,
+        LocalDateTime CanceledAt
+    ): ShoppingCartEvent;
+}
+
+public enum ShoppingCartStatus
+{
+    Pending = 1,
+    Confirmed = 2,
+    Canceled = 4,
+
+    Closed = Confirmed | Canceled
+}
+
+public record PricedProductItem(
+    ProductId ProductId,
+    Quantity Quantity,
+    Price UnitPrice
+);
+
+public record ShoppingCart(
+    ShoppingCartId Id,
+    ClientId ClientId,
+    ShoppingCartStatus Status,
+    Dictionary<ProductId, Quantity> ProductItems
+)
+{
+    public static ShoppingCart When(ShoppingCart entity, object @event)
+    {
+        return @event switch
+        {
+            ShoppingCartOpened (var cartId, var clientId) =>
+                entity with
+                {
+                    Id = cartId,
+                    ClientId = clientId,
+                    Status = ShoppingCartStatus.Pending,
+                    ProductItems = new Dictionary<ProductId, Quantity>()
+                },
+
+            ProductItemAddedToShoppingCart (_, var productItem) =>
+                entity with { ProductItems = entity.ProductItems.Add(productItem) },
+
+            ProductItemRemovedFromShoppingCart (_, var productItem) =>
+                entity with { ProductItems = entity.ProductItems.Remove(productItem) },
+
+            ShoppingCartConfirmed (_, var confirmedAt) =>
+                entity with { Status = ShoppingCartStatus.Confirmed },
+
+            ShoppingCartCanceled (_, var canceledAt) =>
+                entity with { Status = ShoppingCartStatus.Canceled },
+            _ => entity
+        };
+    }
+
+    public static ShoppingCart Default() =>
```

Comment:
never a good pattern. should not be in Domain. Fine if you want it as a helper in tests, but put it there then